### ISSUE TITLE
i64.load{8,16,32}_s only sign-extends to 32 bits

### ISSUE THREAD

**Issue by binji** [WebAssembly/v8-native-prototype#15]
Here's a quick test:
```
(module
  (memory 16 (segment 0 "\ff\ff\ff\ff"))
  (func $test8
    (i64.store (i32.const 8) (i64.load8_s (i32.const 0))))
  (export "test8" $test8)

  (func $test16
    (i64.store (i32.const 8) (i64.load16_s (i32.const 0))))
  (export "test16" $test16)

  (func $test32
    (i64.store (i32.const 8) (i64.load32_s (i32.const 0))))
  (export "test32" $test32)

  (func $get32 (param i32) (result i32)
    (i32.load (get_local 0)))
  (export "get32" $get32))

(invoke "test8")
(invoke "get32" (i32.const 8))
(invoke "get32" (i32.const 12))

(invoke "test16")
(invoke "get32" (i32.const 8))
(invoke "get32" (i32.const 12))

(invoke "test32")
(invoke "get32" (i32.const 8))
(invoke "get32" (i32.const 12))
```

Running this through sexpr-wasm-prototype produces the output:
```
instantiating module
$invoke_0 = undefined
$invoke_1 = -1
$invoke_2 = 0
$invoke_3 = undefined
$invoke_4 = -1
$invoke_5 = 0
$invoke_6 = undefined
$invoke_7 = -1
$invoke_8 = 0
0/0 tests passed.
```

So it looks like it is only sign-extending to 32-bits.

**Comment by titzer** [comment]
Looks like the access byte for load/stores are encoded incorrectly:

; function data 1
0000026: 7e00 0000                                  ; FIXUP func code start offset
000007e: 31                                         ; OPCODE_I64_STORE_I32
000007f: 07                                         ; store access byte
0000080: 10                                         ; OPCODE_I8_CONST
0000081: 08                                         ; u8 literal
0000082: 21                                         ; OPCODE_I64_LOAD_I32
0000083: 05                                         ; load access byte
0000084: 10                                         ; OPCODE_I8_CONST
0000085: 00                                         ; u8 literal
000002a: 8600 0000                                  ; FIXUP func code end offset
0000036: 01                                         ; FIXUP func exported

Here OPCODE_I64_LOAD_I32 has load access byte 05, which actually encodes i64.load16_s.

The meaning of access byte is defined in src/wasm/wasm-opcodes.h:

// Functionality related to encoding memory accesses.                                                                                                                      
struct MemoryAccess {
  // Atomicity annotations for access to the memory and globals.                                                                                                           
  enum Atomicity {
    kNone = 0,        // non-atomic                                                                                                                                        
    kSequential = 1,  // sequential consistency                                                                                                                            
    kAcquire = 2,     // acquire semantics                                                                                                                                 
    kRelease = 3      // release semantics                                                                                                                                 
  };

  // Alignment annotations for memory accesses.                                                                                                                            
  enum Alignment { kAligned = 0, kUnaligned = 1 };

  // Memory width for integer accesses.                                                                                                                                    
  enum IntWidth { kI8 = 0, kI16 = 1, kI32 = 2, kI64 = 3 };

  // Bitfields for the various annotations for memory accesses.                                                                                                            
  typedef BitField<IntWidth, 0, 2> IntWidthField;
  typedef BitField<bool, 2, 1> SignExtendField;
  typedef BitField<Alignment, 3, 1> AlignmentField;
  typedef BitField<Atomicity, 4, 2> AtomicityField;
};

So, the bit layout looks like this:
bits[0...1] = integer width (i.e. memory size)
bits[2..2] = sign extend
bits[3..3] = alignment
bits[4..5] = atomicity

It's a little hard to read because of the usage of the V8 BitField class.

**Comment by binji** [comment]
Yeah, I was read that file originally to see how it works:
this is what I have currently in hash.txt, the last value is the memory access byte:
```
i64.load16_s, WASM_OP_LOAD, WASM_OPCODE_I64_LOAD_I32, WASM_TYPE_I64, WASM_MEM_TYPE_I16, 5
i64.load16_u, WASM_OP_LOAD, WASM_OPCODE_I64_LOAD_I32, WASM_TYPE_I64, WASM_MEM_TYPE_I16, 1
i64.load32_s, WASM_OP_LOAD, WASM_OPCODE_I64_LOAD_I32, WASM_TYPE_I64, WASM_MEM_TYPE_I32, 6
i64.load32_u, WASM_OP_LOAD, WASM_OPCODE_I64_LOAD_I32, WASM_TYPE_I64, WASM_MEM_TYPE_I32, 2
i64.load8_s, WASM_OP_LOAD, WASM_OPCODE_I64_LOAD_I32, WASM_TYPE_I64, WASM_MEM_TYPE_I8, 4
i64.load8_u, WASM_OP_LOAD, WASM_OPCODE_I64_LOAD_I32, WASM_TYPE_I64, WASM_MEM_TYPE_I8, 0
i64.load, WASM_OP_LOAD, WASM_OPCODE_I64_LOAD_I32, WASM_TYPE_I64, WASM_MEM_TYPE_I64, 7
```
It looks right to me, and all the other loads work properly (i32 of all sizes and i64 unsigned). Only i64 signed loads are failing.

**Comment by titzer** [comment]
I'll look again. I obviously have some tests for i64 that are at a lower
level, so it's not fundamentally broken :-) How do you get (invoke)
commands to work properly? I get parse errors.

**Comment by binji** [comment]
Yeah, I should probably invert those flags so invoke works by default.
`test/run-d8.py --spec test.wasm` to run spec tests.
to see the actual output: `out/sexpr-wasm --multi-module test.wasm -o test.bin`

**Comment by binji** [comment]
Fixed by 1e3d7a199848a4850085e5c7f9dcb524d3f23d6a.
